### PR TITLE
Adding an automated password reset option

### LINKED ISSUE #621: Automated password reset
Via email

### PR DESCRIPTION
**Pull request by kev306**
Resolves #621

### CODE REVIEW

**Comment by kev306** [comment]
In progress still

**Comment by kev306** [comment]
Reset Password page

image

File: src/myFunctions/emailTemplateResetPassword.ts [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+                      <table class="main">
+                        <tr>
+                          <td class="wrapper">
+                            <table>
+                              <tr>
+                                <td>
+                                  <p>Hey!</p></br>
+                                  <p>You have recently requested to reset your password. To complete this process, please click the following link. Note that the link will expire after 1 hour.</p></br>
+                                  <p><a href="http://<%=server_domain%>/resetPassword/verifyResetPassword?token=<%=token%>">http://<%=server_domain%>/resetPassword/verifyResetPassword?token=<%=token%></a></p>
+                                  <p>If this was not you, please let us know immediately.</p>
```
Comment (vck3000) [inline src/myFunctions/emailTemplateResetPassword.ts:436]:
```suggestion
                                  <p>If this was not you, you may disregard this email.</p>
```

File: src/views/resetPassword.ejs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                    class="form-control"
+                    placeholder="Email Address"
+                    required
+            />
+            <div class="help-block with-errors"></div>
+        </div>
+
+        <br />
+
+        <% if(locals.platform && platform==="prod" ){ %>
```
Comment (vck3000) [inline src/views/resetPassword.ejs:51]:
```suggestion
        <% if(locals.platform && platform === "prod" ){ %>

```

Comment (vck3000) [inline src/views/resetPassword.ejs:51]: Can you try without `locals.platform`?

File: src/routes/index.js [hunk 3]
```diff
@@ -247,6 +249,120 @@ router.get('/statistics', (req, res) => {
...
+      emailVerified: true,
+    });
+
+    if (user) {
+      sendResetPassword(user, email);
+      req.flash(
+        'success',
+        'A link to reset your password has been sent to your email.',
+      );
+      res.redirect('/resetPassword');
```
Comment (vck3000) [inline src/routes/index.js:306]:
```suggestion
      res.redirect('/');
```

File: src/routes/index.js [hunk 4]
```diff
@@ -247,6 +249,120 @@ router.get('/statistics', (req, res) => {
...
+      user &&
+      user.emailTokenExpiry &&
+      new Date().getTime() < user.emailTokenExpiry.getTime()
+    ) {
+      user.emailToken = undefined;
+      user.markModified('emailToken');
+
+      // Set new temporary password
+      const uuidv4 = uuid.v4;
+      const newPassword = uuidv4().substring(0, 12);
```
Comment (vck3000) [inline src/routes/index.js:341]: Can we just inline this? 

```
const newPassword = uuid.v4().substring(0, 12);
```

File: src/routes/index.js [hunk 5]
```diff
@@ -247,6 +249,120 @@ router.get('/statistics', (req, res) => {
...
+          if (err) {
+            reject(err);
+          }
+          resolve();
+        });
+      });
+
+      await user.save();
+
+      req.flash('success', 'Your password has been reset! Thank you!');
```
Comment (vck3000) [inline src/routes/index.js:354]:
```suggestion
      req.flash('success', 'Your password has been reset!');
```

File: src/views/resetPasswordSuccess.ejs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    }
+    .reset-password h2 {
+        margin: 0 0 15px;
+    }
+</style>
+
+<div class="reset-password">
+    <h2 class="text-center">Your password has been reset!</h2>
+    <br>
+    <p> Here is your new password: <b><%=newPassword%></b></p>
```
Comment (vck3000) [inline src/views/resetPasswordSuccess.ejs:21]:
```suggestion
    <p> Your new password is: <b><%=newPassword%></b></p>

```

File: src/views/resetPasswordSuccess.ejs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    <br>
+    <p> Please change your password when able to.</p>
```
Comment (vck3000) [inline src/views/resetPasswordSuccess.ejs:23]:
```suggestion
    <p> Please change your password when you log on.</p>

```

File: src/views/troubleshooting.ejs [hunk 8]
```diff
@@ -30,7 +30,7 @@
 
   <ul>
     <li class="question">Question: I need to reset my password!</li>
-    <li class="answer">Contact me directly through discord.</li>
+    <li class="answer">You can reset your password <a href="/resetPassword">here</a>.</li>
     <li class="question">Question: I need to change my password!</li>
     <li class="answer">You can change your password within your profile.</li>
     <li class="question">Question: Nothing is loading!</li>
```
Comment (vck3000) [inline src/views/troubleshooting.ejs:36]: Let's just remove troubleshooting.ejs lol

File: src/views/resetPasswordSuccess.ejs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    <p>If you are still running into issues, please check out our <a href="/troubleshooting">troubleshooting page</a>.</p>
```
Comment (vck3000) [inline src/views/resetPasswordSuccess.ejs:24]: Let's remove this line

**Review by vck3000 — CHANGES_REQUESTED** [review]
Nice! Give it a test and please fix up the small nits :)

File: src/myFunctions/sendResetPassword.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+const uuidv4 = uuid.v4;
+
+export const sendResetPassword = async (user: any, email: string) => {
+  const token = uuidv4();
+  const currentDate = new Date();
+  let tokenDateExpired = new Date();
+
+  // Set tokenDateExpired to be 1 hour later
+  tokenDateExpired.setTime(currentDate.getTime() + TOKEN_TIMEOUT);
```
Comment (vck3000) [inline src/myFunctions/sendResetPassword.ts:18]: Can combine this with the above declaration line

File: src/views/troubleshooting.ejs [hunk 11]
```diff
@@ -30,7 +30,7 @@
 
   <ul>
     <li class="question">Question: I need to reset my password!</li>
-    <li class="answer">Contact me directly through discord.</li>
+    <li class="answer">You can reset your password <a href="/resetPassword">here</a>.</li>
     <li class="question">Question: I need to change my password!</li>
```
Comment (vck3000) [inline src/views/troubleshooting.ejs:34]: Let’s delete this file if we’ve removed the route. Don’t think we need it anymore.

File: src/myFunctions/sendEmail.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+const api_key = process.env.MAILGUN_API_KEY;
+const domain = process.env.PROAVALON_EMAIL_ADDRESS_DOMAIN;
+
+const mailgun = new Mailgun(formData);
+const mg = mailgun.client({ username: 'api', key: api_key });
+
+export const sendEmail = (
+  recipientEmail: string,
+  subject: string,
+  message: string,
```
Comment (vck3000) [inline src/myFunctions/sendEmail.ts:13]: Maybe can call this `messageHtml`

File: src/myFunctions/sendEmailVerification.ts [hunk 13]
```diff
@@ -1,17 +1,11 @@
...
-import Mailgun from 'mailgun.js';
-import emailTemplate from './emailTemplate';
+import emailTemplateEmailVerification from './emailTemplateEmailVerification';
 import uuid from 'uuid';
 import disposableEmails from '../util/disposableEmails.js';
+import { sendEmail } from './sendEmail';
 
-const api_key = process.env.MAILGUN_API_KEY;
-const domain = process.env.PROAVALON_EMAIL_ADDRESS_DOMAIN;
 const server_domain = process.env.SERVER_DOMAIN;
```
Comment (vck3000) [inline src/myFunctions/sendEmailVerification.ts:7]: Let's camel case this `serverDomain`

File: src/myFunctions/sendResetPassword.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
+import uuid from 'uuid';
+import ejs from 'ejs';
+import emailTemplateResetPassword from './emailTemplateResetPassword';
+import { sendEmail } from './sendEmail';
+
+const TOKEN_TIMEOUT = 60 * 60 * 1000; // 1 hour
+
+const server_domain = process.env.SERVER_DOMAIN;
```
Comment (vck3000) [inline src/myFunctions/sendResetPassword.ts:8]: Same here

File: src/myFunctions/sendResetPassword.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+const uuidv4 = uuid.v4;
```
Comment (vck3000) [inline src/myFunctions/sendResetPassword.ts:10]: Combine this guy

File: src/myFunctions/sendResetPassword.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+export const sendResetPassword = async (user: any, email: string) => {
```
Comment (vck3000) [inline src/myFunctions/sendResetPassword.ts:12]: Maybe change the `any` to `IUser` if you're bothered lol

File: src/myFunctions/sendResetPassword.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  const token = uuidv4();
+  const currentDate = new Date();
+
+  // Set tokenDateExpired to be 1 hour later
+  let tokenDateExpired = new Date().setTime(
```
Comment (vck3000) [inline src/myFunctions/sendResetPassword.ts:17]: Let's double check this is correct.

Comment (vck3000) [reply src/myFunctions/sendResetPassword.ts:17]: Can remove setTime()?

File: src/routes/index.js [hunk 18]
```diff
@@ -239,14 +241,123 @@ router.get('/security', (req, res) => {
...
+    if (body.success !== undefined && !body.success) {
+      req.flash('error', 'Failed captcha verification.');
+      res.redirect('/resetPassword');
+      return;
+    }
+  }
+
+  const secretKey = process.env.MY_SECRET_GOOGLE_CAPTCHA_KEY;
+
+  const verifyUrl = `https://google.com/recaptcha/api/siteverify?secret=${secretKey}&response=${req.body.captcha}&remoteip=${req.connection.remoteAddress}`;
```
Comment (vck3000) [inline src/routes/index.js:280]: Would be nice to extract this out to reduce duplication but let's leave it for now so you can get your change out sooner.

File: src/routes/index.js [hunk 19]
```diff
@@ -239,14 +241,123 @@ router.get('/security', (req, res) => {
...
+      sendResetPassword(user, email);
+      req.flash(
+        'success',
+        'A link to reset your password has been sent to your email.',
+      );
+      res.redirect('/');
+      console.log(
+        `User: ${user.username} Email: ${user.emailAddress} has requested to reset their password.`,
+      );
+    } else {
```
Comment (vck3000) [inline src/routes/index.js:306]: Let's flip this guy and early return if no user

File: src/routes/index.js [hunk 20]
```diff
@@ -239,14 +241,123 @@ router.get('/security', (req, res) => {
...
+      req.flash(
+        'error',
+        'Email does not exist. Please enter a registered email address.',
+      );
+      res.redirect('/resetPassword');
+      console.log(`Email not found: ${email}`);
+    }
+  } catch (error) {
```
Comment (vck3000) [inline src/routes/index.js:314]: I don't think we need this. Express should handle this for us if we throw an error.

File: src/routes/index.js [hunk 21]
```diff
@@ -239,14 +241,123 @@ router.get('/security', (req, res) => {
...
+      });
+
+      await user.save();
+
+      req.flash('success', 'Your password has been reset!');
+      res.render('resetPasswordSuccess', { newPassword });
+      return;
+    }
+  }
+  req.flash(
```
Comment (vck3000) [inline src/routes/index.js:354]: Nit: new line above

File: src/views/resetPassword.ejs [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+    .btn {
+        font-size: 15px;
+        font-weight: bold;
+    }
+</style>
+
+<div class="login-form">
+    <form action="/resetPassword" method="POST" id="resetPasswordId" data-toggle="validator">
+
+        <h2 class="text-center">Let's reset your password!</h2>
```
Comment (vck3000) [inline src/views/resetPassword.ejs:34]: `Reset Password`?

**Review by vck3000 — APPROVED** [review]
Changelog!

File: src/gameplay/types.ts [hunk 23]
```diff
@@ -40,6 +40,7 @@ export interface IUser {
   emailAddress?: string;
```
Comment (vck3000) [inline src/gameplay/types.ts:40]: Changelog!

File: src/myFunctions/sendResetPassword.ts (hunk 16 above)
Comment (kev306) [reply src/myFunctions/sendResetPassword.ts:10]: Will look into this in a future PR

**Pull request merged**